com.hazelcast.cluster.TcpIpJoinTest

_master_ (commit c3c641fc6af0924ccc140f24d98e37f7f8dd66b1)

Failed on IBM JDK 8: http://jenkins.hazelcast.com/view/Official%20Builds/job/Hazelcast-master-IbmJDK8-nightly/1/testReport/com.hazelcast.cluster/TcpIpJoinTest/

Following tests from `com.hazelcast.cluster.TcpIpJoinTest` failed with similar exception:
- `test_whenNoExplicitPortConfigured`
- `test_whenHostAndInterfacesConfigured`
- `test_whenPortAndInterfacesConfigured`
- `test_whenExplicitPortConfigured`

Stacktrace (for `test_whenNoExplicitPortConfigured`):
```
java.lang.AssertionError: Cluster size is not correct. Expected: 2, actual: 1, instance index: 0
	at org.junit.Assert.fail(Assert.java:89)
	at com.hazelcast.test.HazelcastTestSupport.assertClusterSize(HazelcastTestSupport.java:1035)
	at com.hazelcast.cluster.AbstractJoinTest.testJoin(AbstractJoinTest.java:49)
	at com.hazelcast.cluster.TcpIpJoinTest.test_whenNoExplicitPortConfigured(TcpIpJoinTest.java:60)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:90)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:55)
	at java.lang.reflect.Method.invoke(Method.java:508)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:115)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:107)
	at java.util.concurrent.FutureTask.run(FutureTask.java:277)
	at java.lang.Thread.run(Thread.java:822)
```

Standard output (for `test_whenNoExplicitPortConfigured`):
```
09:01:32,089  INFO |test_whenNoExplicitPortConfigured| - [MetricsConfigHelper] test_whenNoExplicitPortConfigured - [LOCAL] [dev] [5.0-SNAPSHOT] Overridden metrics configuration with system property 'hazelcast.metrics.collection.frequency'='1' -> 'MetricsConfig.collectionFrequencySeconds'='1'
09:01:32,090  INFO |test_whenNoExplicitPortConfigured| - [system] test_whenNoExplicitPortConfigured - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] 
	o  o   O   o---o o--o o      o-o   O    o-o  o-O-o     o--o  o       O  o-O-o o--o  o-o  o--o  o   o 
	|  |  / \     /  |    |     /     / \  |       |       |   | |      / \   |   |    o   o |   | |\ /| 
	O--O o---o  -O-  O-o  |    O     o---o  o-o    |       O--o  |     o---o  |   O-o  |   | O-Oo  | O | 
	|  | |   |  /    |    |     \    |   |     |   |       |     |     |   |  |   |    o   o |  \  |   | 
	o  o o   o o---o o--o O---o  o-o o   o o--o    o       o     O---o o   o  o   o     o-o  o   o o   o
09:01:32,090  INFO |test_whenNoExplicitPortConfigured| - [system] test_whenNoExplicitPortConfigured - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Copyright (c) 2008-2021, Hazelcast, Inc. All Rights Reserved.
09:01:32,090  INFO |test_whenNoExplicitPortConfigured| - [system] test_whenNoExplicitPortConfigured - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Hazelcast Platform 5.0-SNAPSHOT (20210524 - c3c641f) starting at [127.0.0.1]:5701
09:01:32,090  INFO |test_whenNoExplicitPortConfigured| - [system] test_whenNoExplicitPortConfigured - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Cluster name: dev
09:01:32,092  INFO |test_whenNoExplicitPortConfigured| - [MetricsConfigHelper] test_whenNoExplicitPortConfigured - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Collecting debug metrics and sending to diagnostics is enabled
09:01:32,097  INFO |test_whenNoExplicitPortConfigured| - [Node] test_whenNoExplicitPortConfigured - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Using TCP/IP discovery
09:01:32,097  WARN |test_whenNoExplicitPortConfigured| - [CPSubsystem] test_whenNoExplicitPortConfigured - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] CP Subsystem is not enabled. CP data structures will operate in UNSAFE mode! Please note that UNSAFE mode will not provide strong consistency guarantees.
09:01:32,112  INFO |test_whenNoExplicitPortConfigured| - [JetService] test_whenNoExplicitPortConfigured - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Setting number of cooperative threads and default parallelism to 72
09:01:32,131  INFO |test_whenNoExplicitPortConfigured| - [Diagnostics] test_whenNoExplicitPortConfigured - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Diagnostics disabled. To enable add -Dhazelcast.diagnostics.enabled=true to the JVM arguments.
09:01:32,131  INFO |test_whenNoExplicitPortConfigured| - [LifecycleService] test_whenNoExplicitPortConfigured - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] [127.0.0.1]:5701 is STARTING
09:01:32,135  INFO |test_whenNoExplicitPortConfigured| - [TcpIpJoiner] hz.friendly_blackburn.cached.thread-2 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] [127.0.0.1]:5702 is added to the blacklist.
09:01:32,135  INFO |test_whenNoExplicitPortConfigured| - [TcpIpJoiner] hz.friendly_blackburn.cached.thread-3 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] [127.0.0.1]:5703 is added to the blacklist.
09:01:32,235  INFO |test_whenNoExplicitPortConfigured| - [TcpIpJoiner] hz.friendly_blackburn.cached.thread-3 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] [127.0.0.1]:5703 is added to the blacklist.
09:01:32,235  INFO |test_whenNoExplicitPortConfigured| - [TcpIpJoiner] hz.friendly_blackburn.cached.thread-2 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] [127.0.0.1]:5702 is added to the blacklist.
09:01:32,435  INFO |test_whenNoExplicitPortConfigured| - [TcpIpJoiner] hz.friendly_blackburn.cached.thread-3 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] [127.0.0.1]:5702 is added to the blacklist.
09:01:32,435  INFO |test_whenNoExplicitPortConfigured| - [TcpIpJoiner] hz.friendly_blackburn.cached.thread-2 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] [127.0.0.1]:5703 is added to the blacklist.
09:01:32,735  INFO |test_whenNoExplicitPortConfigured| - [TcpIpJoiner] hz.friendly_blackburn.cached.thread-3 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] [127.0.0.1]:5702 is added to the blacklist.
09:01:32,735  INFO |test_whenNoExplicitPortConfigured| - [TcpIpJoiner] hz.friendly_blackburn.cached.thread-1 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] [127.0.0.1]:5703 is added to the blacklist.
09:01:33,135  INFO |test_whenNoExplicitPortConfigured| - [JetExtension] test_whenNoExplicitPortConfigured - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Jet extension is enabled after the cluster version upgrade.
09:01:33,135  INFO |test_whenNoExplicitPortConfigured| - [ClusterService] test_whenNoExplicitPortConfigured - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] 

Members {size:1, ver:1} [
	Member [127.0.0.1]:5701 - 70b81027-7809-49bb-8895-e4adf0fc7475 this
]

09:01:33,135  INFO |test_whenNoExplicitPortConfigured| - [JetExtension] test_whenNoExplicitPortConfigured - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Jet extension is enabled
09:01:33,135 DEBUG |test_whenNoExplicitPortConfigured| - [JobCoordinationService] hz.friendly_blackburn.cached.thread-2 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Not starting jobs because partitions are not yet initialized.
09:01:33,135 DEBUG |test_whenNoExplicitPortConfigured| - [JobCoordinationService] hz.friendly_blackburn.cached.thread-5 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Not starting jobs because partitions are not yet initialized.
09:01:33,135  INFO |test_whenNoExplicitPortConfigured| - [LifecycleService] test_whenNoExplicitPortConfigured - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] [127.0.0.1]:5701 is STARTED
09:01:33,135  INFO |test_whenNoExplicitPortConfigured| - [MetricsConfigHelper] test_whenNoExplicitPortConfigured - [LOCAL] [dev] [5.0-SNAPSHOT] Overridden metrics configuration with system property 'hazelcast.metrics.collection.frequency'='1' -> 'MetricsConfig.collectionFrequencySeconds'='1'
09:01:33,136  INFO |test_whenNoExplicitPortConfigured| - [HealthMonitor] hz.friendly_blackburn.HealthMonitor - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] processors=72, physical.memory.total=377.6G, physical.memory.free=151.3G, swap.space.total=4.0G, swap.space.free=2.2G, heap.memory.used=2.9G, heap.memory.free=1.1G, heap.memory.total=4.0G, heap.memory.max=4.0G, heap.memory.used/total=72.96%, heap.memory.used/max=72.96%, minor.gc.count=10599, minor.gc.time=275614ms, major.gc.count=67, major.gc.time=25273ms, load.process=1.89%, load.system=11.49%, load.systemAverage=5.81, thread.count=393, thread.peakCount=12615, cluster.timeDiff=0, event.q.size=0, executor.q.async.size=0, executor.q.client.size=0, executor.q.client.query.size=0, executor.q.client.blocking.size=0, executor.q.query.size=0, executor.q.scheduled.size=0, executor.q.io.size=0, executor.q.system.size=0, executor.q.operations.size=0, executor.q.priorityOperation.size=0, operations.completed.count=0, executor.q.mapLoad.size=0, executor.q.mapLoadAllKeys.size=0, executor.q.cluster.size=0, executor.q.response.size=0, operations.running.count=0, operations.pending.invocations.percentage=0.00%, operations.pending.invocations.count=0, proxy.count=0, clientEndpoint.count=0, connection.active.count=0, client.connection.count=0, connection.count=0
09:01:33,137  INFO |test_whenNoExplicitPortConfigured| - [system] test_whenNoExplicitPortConfigured - [127.0.0.1]:5702 [dev] [5.0-SNAPSHOT] 
	o  o   O   o---o o--o o      o-o   O    o-o  o-O-o     o--o  o       O  o-O-o o--o  o-o  o--o  o   o 
	|  |  / \     /  |    |     /     / \  |       |       |   | |      / \   |   |    o   o |   | |\ /| 
	O--O o---o  -O-  O-o  |    O     o---o  o-o    |       O--o  |     o---o  |   O-o  |   | O-Oo  | O | 
	|  | |   |  /    |    |     \    |   |     |   |       |     |     |   |  |   |    o   o |  \  |   | 
	o  o o   o o---o o--o O---o  o-o o   o o--o    o       o     O---o o   o  o   o     o-o  o   o o   o
09:01:33,137  INFO |test_whenNoExplicitPortConfigured| - [system] test_whenNoExplicitPortConfigured - [127.0.0.1]:5702 [dev] [5.0-SNAPSHOT] Copyright (c) 2008-2021, Hazelcast, Inc. All Rights Reserved.
09:01:33,137  INFO |test_whenNoExplicitPortConfigured| - [system] test_whenNoExplicitPortConfigured - [127.0.0.1]:5702 [dev] [5.0-SNAPSHOT] Hazelcast Platform 5.0-SNAPSHOT (20210524 - c3c641f) starting at [127.0.0.1]:5702
09:01:33,137  INFO |test_whenNoExplicitPortConfigured| - [system] test_whenNoExplicitPortConfigured - [127.0.0.1]:5702 [dev] [5.0-SNAPSHOT] Cluster name: dev
09:01:33,139  INFO |test_whenNoExplicitPortConfigured| - [MetricsConfigHelper] test_whenNoExplicitPortConfigured - [127.0.0.1]:5702 [dev] [5.0-SNAPSHOT] Collecting debug metrics and sending to diagnostics is enabled
09:01:33,144  INFO |test_whenNoExplicitPortConfigured| - [Node] test_whenNoExplicitPortConfigured - [127.0.0.1]:5702 [dev] [5.0-SNAPSHOT] Using TCP/IP discovery
09:01:33,144  WARN |test_whenNoExplicitPortConfigured| - [CPSubsystem] test_whenNoExplicitPortConfigured - [127.0.0.1]:5702 [dev] [5.0-SNAPSHOT] CP Subsystem is not enabled. CP data structures will operate in UNSAFE mode! Please note that UNSAFE mode will not provide strong consistency guarantees.
09:01:33,163  INFO |test_whenNoExplicitPortConfigured| - [JetService] test_whenNoExplicitPortConfigured - [127.0.0.1]:5702 [dev] [5.0-SNAPSHOT] Setting number of cooperative threads and default parallelism to 72
09:01:33,186  INFO |test_whenNoExplicitPortConfigured| - [Diagnostics] test_whenNoExplicitPortConfigured - [127.0.0.1]:5702 [dev] [5.0-SNAPSHOT] Diagnostics disabled. To enable add -Dhazelcast.diagnostics.enabled=true to the JVM arguments.
09:01:33,186  INFO |test_whenNoExplicitPortConfigured| - [LifecycleService] test_whenNoExplicitPortConfigured - [127.0.0.1]:5702 [dev] [5.0-SNAPSHOT] [127.0.0.1]:5702 is STARTING
09:01:33,190  INFO |test_whenNoExplicitPortConfigured| - [TcpServerConnection] hz.friendly_blackburn.IO.thread-in-2 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Initialized new cluster connection between /127.0.0.1:5701 and /127.0.0.1:56585
09:01:33,190  INFO |test_whenNoExplicitPortConfigured| - [TcpIpJoiner] hz.stupefied_blackburn.cached.thread-2 - [127.0.0.1]:5702 [dev] [5.0-SNAPSHOT] [127.0.0.1]:5703 is added to the blacklist.
09:01:33,190  INFO |test_whenNoExplicitPortConfigured| - [TcpServerConnection] hz.stupefied_blackburn.IO.thread-in-1 - [127.0.0.1]:5702 [dev] [5.0-SNAPSHOT] Initialized new cluster connection between /0.0.0.0:56585 and /127.0.0.1:5701
09:01:33,235 DEBUG |test_whenNoExplicitPortConfigured| - [JobCoordinationService] hz.friendly_blackburn.cached.thread-5 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Not starting jobs because partitions are not yet initialized.
09:01:33,235 DEBUG |test_whenNoExplicitPortConfigured| - [JobCoordinationService] hz.friendly_blackburn.cached.thread-5 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Not starting jobs because partitions are not yet initialized.
09:01:33,289  INFO |test_whenNoExplicitPortConfigured| - [TcpIpJoiner] hz.stupefied_blackburn.cached.thread-2 - [127.0.0.1]:5702 [dev] [5.0-SNAPSHOT] [127.0.0.1]:5703 is added to the blacklist.
09:01:33,291  INFO |test_whenNoExplicitPortConfigured| - [ClusterService] hz.friendly_blackburn.priority-generic-operation.thread-0 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] 

Members {size:2, ver:2} [
	Member [127.0.0.1]:5701 - 70b81027-7809-49bb-8895-e4adf0fc7475 this
	Member [127.0.0.1]:5702 - bce25553-6b6d-4d14-a003-674b088f9bc6
]

09:01:33,291  INFO |test_whenNoExplicitPortConfigured| - [JetExtension] hz.stupefied_blackburn.priority-generic-operation.thread-0 - [127.0.0.1]:5702 [dev] [5.0-SNAPSHOT] Jet extension is enabled after the cluster version upgrade.
09:01:33,291  INFO |test_whenNoExplicitPortConfigured| - [ClusterService] hz.stupefied_blackburn.priority-generic-operation.thread-0 - [127.0.0.1]:5702 [dev] [5.0-SNAPSHOT] 

Members {size:2, ver:2} [
	Member [127.0.0.1]:5701 - 70b81027-7809-49bb-8895-e4adf0fc7475
	Member [127.0.0.1]:5702 - bce25553-6b6d-4d14-a003-674b088f9bc6 this
]

09:01:33,291  INFO |test_whenNoExplicitPortConfigured| - [JetExtension] test_whenNoExplicitPortConfigured - [127.0.0.1]:5702 [dev] [5.0-SNAPSHOT] Jet extension is enabled
09:01:33,292  INFO |test_whenNoExplicitPortConfigured| - [LifecycleService] test_whenNoExplicitPortConfigured - [127.0.0.1]:5702 [dev] [5.0-SNAPSHOT] [127.0.0.1]:5702 is STARTED
09:01:33,292  INFO |test_whenNoExplicitPortConfigured| - [LifecycleService] test_whenNoExplicitPortConfigured - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] [127.0.0.1]:5701 is SHUTTING_DOWN
09:01:33,292 DEBUG |test_whenNoExplicitPortConfigured| - [JobCoordinationService] test_whenNoExplicitPortConfigured - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Added a shutting-down member: 70b81027-7809-49bb-8895-e4adf0fc7475
09:01:33,292  INFO |test_whenNoExplicitPortConfigured| - [Node] test_whenNoExplicitPortConfigured - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Shutting down connection manager...
09:01:33,292  INFO |test_whenNoExplicitPortConfigured| - [TcpServerConnection] test_whenNoExplicitPortConfigured - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Connection[id=1, /127.0.0.1:5701->/127.0.0.1:56585, qualifier=null, endpoint=[127.0.0.1]:5702, alive=false, connectionType=MEMBER, planeIndex=0] closed. Reason: TcpServer is stopping
09:01:33,292  INFO |test_whenNoExplicitPortConfigured| - [TcpIpJoiner] test_whenNoExplicitPortConfigured - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] [127.0.0.1]:5702 is added to the blacklist.
09:01:33,293  INFO |test_whenNoExplicitPortConfigured| - [TcpServerConnection] hz.stupefied_blackburn.IO.thread-in-1 - [127.0.0.1]:5702 [dev] [5.0-SNAPSHOT] Connection[id=1, /0.0.0.0:56585->/127.0.0.1:5701, qualifier=null, endpoint=[127.0.0.1]:5701, alive=false, connectionType=MEMBER, planeIndex=0] closed. Reason: Connection closed by the other side
09:01:33,293  INFO |test_whenNoExplicitPortConfigured| - [Node] test_whenNoExplicitPortConfigured - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Shutting down node engine...
09:01:33,293  INFO |test_whenNoExplicitPortConfigured| - [TcpServerConnector] hz.stupefied_blackburn.cached.thread-4 - [127.0.0.1]:5702 [dev] [5.0-SNAPSHOT] Connecting to /127.0.0.1:5701, timeout: 10000, bind-any: true
09:01:33,293  INFO |test_whenNoExplicitPortConfigured| - [HealthMonitor] hz.stupefied_blackburn.HealthMonitor - [127.0.0.1]:5702 [dev] [5.0-SNAPSHOT] processors=72, physical.memory.total=377.6G, physical.memory.free=151.2G, swap.space.total=4.0G, swap.space.free=2.2G, heap.memory.used=2.9G, heap.memory.free=1.1G, heap.memory.total=4.0G, heap.memory.max=4.0G, heap.memory.used/total=73.14%, heap.memory.used/max=73.14%, minor.gc.count=10599, minor.gc.time=275614ms, major.gc.count=67, major.gc.time=25273ms, load.process=2.91%, load.system=16.04%, load.systemAverage=5.81, thread.count=598, thread.peakCount=12615, cluster.timeDiff=0, event.q.size=0, executor.q.async.size=0, executor.q.client.size=0, executor.q.client.query.size=0, executor.q.client.blocking.size=0, executor.q.query.size=0, executor.q.scheduled.size=0, executor.q.io.size=0, executor.q.system.size=0, executor.q.operations.size=0, executor.q.priorityOperation.size=0, operations.completed.count=12, executor.q.mapLoad.size=0, executor.q.mapLoadAllKeys.size=0, executor.q.cluster.size=0, executor.q.response.size=0, operations.running.count=0, operations.pending.invocations.percentage=0.00%, operations.pending.invocations.count=0, proxy.count=0, clientEndpoint.count=0, connection.active.count=0, client.connection.count=0, connection.count=0
09:01:33,293  INFO |test_whenNoExplicitPortConfigured| - [TcpServerConnector] hz.stupefied_blackburn.cached.thread-4 - [127.0.0.1]:5702 [dev] [5.0-SNAPSHOT] Could not connect to: /127.0.0.1:5701. Reason: IOException[Connection refused to address /127.0.0.1:5701]
09:01:33,299  INFO |test_whenNoExplicitPortConfigured| - [NodeExtension] test_whenNoExplicitPortConfigured - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Destroying node NodeExtension.
09:01:33,299  INFO |test_whenNoExplicitPortConfigured| - [Node] test_whenNoExplicitPortConfigured - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Hazelcast Shutdown is completed in 7 ms.
09:01:33,299  INFO |test_whenNoExplicitPortConfigured| - [LifecycleService] test_whenNoExplicitPortConfigured - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] [127.0.0.1]:5701 is SHUTDOWN
09:01:33,299  INFO |test_whenNoExplicitPortConfigured| - [MetricsConfigHelper] test_whenNoExplicitPortConfigured - [LOCAL] [dev] [5.0-SNAPSHOT] Overridden metrics configuration with system property 'hazelcast.metrics.collection.frequency'='1' -> 'MetricsConfig.collectionFrequencySeconds'='1'
09:01:33,301  INFO |test_whenNoExplicitPortConfigured| - [system] test_whenNoExplicitPortConfigured - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] 
	o  o   O   o---o o--o o      o-o   O    o-o  o-O-o     o--o  o       O  o-O-o o--o  o-o  o--o  o   o 
	|  |  / \     /  |    |     /     / \  |       |       |   | |      / \   |   |    o   o |   | |\ /| 
	O--O o---o  -O-  O-o  |    O     o---o  o-o    |       O--o  |     o---o  |   O-o  |   | O-Oo  | O | 
	|  | |   |  /    |    |     \    |   |     |   |       |     |     |   |  |   |    o   o |  \  |   | 
	o  o o   o o---o o--o O---o  o-o o   o o--o    o       o     O---o o   o  o   o     o-o  o   o o   o
09:01:33,301  INFO |test_whenNoExplicitPortConfigured| - [system] test_whenNoExplicitPortConfigured - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Copyright (c) 2008-2021, Hazelcast, Inc. All Rights Reserved.
09:01:33,301  INFO |test_whenNoExplicitPortConfigured| - [system] test_whenNoExplicitPortConfigured - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Hazelcast Platform 5.0-SNAPSHOT (20210524 - c3c641f) starting at [127.0.0.1]:5701
09:01:33,301  INFO |test_whenNoExplicitPortConfigured| - [system] test_whenNoExplicitPortConfigured - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Cluster name: dev
09:01:33,303  INFO |test_whenNoExplicitPortConfigured| - [MetricsConfigHelper] test_whenNoExplicitPortConfigured - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Collecting debug metrics and sending to diagnostics is enabled
09:01:33,309  INFO |test_whenNoExplicitPortConfigured| - [Node] test_whenNoExplicitPortConfigured - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Using TCP/IP discovery
09:01:33,309  WARN |test_whenNoExplicitPortConfigured| - [CPSubsystem] test_whenNoExplicitPortConfigured - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] CP Subsystem is not enabled. CP data structures will operate in UNSAFE mode! Please note that UNSAFE mode will not provide strong consistency guarantees.
09:01:33,325  INFO |test_whenNoExplicitPortConfigured| - [JetService] test_whenNoExplicitPortConfigured - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Setting number of cooperative threads and default parallelism to 72
09:01:33,341  INFO |test_whenNoExplicitPortConfigured| - [Diagnostics] test_whenNoExplicitPortConfigured - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Diagnostics disabled. To enable add -Dhazelcast.diagnostics.enabled=true to the JVM arguments.
09:01:33,341  INFO |test_whenNoExplicitPortConfigured| - [LifecycleService] test_whenNoExplicitPortConfigured - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] [127.0.0.1]:5701 is STARTING
09:01:33,344  INFO |test_whenNoExplicitPortConfigured| - [TcpIpJoiner] hz.dreamy_blackburn.cached.thread-2 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] [127.0.0.1]:5703 is added to the blacklist.
09:01:33,344  INFO |test_whenNoExplicitPortConfigured| - [TcpServerConnection] hz.stupefied_blackburn.IO.thread-in-0 - [127.0.0.1]:5702 [dev] [5.0-SNAPSHOT] Initialized new cluster connection between /127.0.0.1:5702 and /127.0.0.1:40592
09:01:33,345  INFO |test_whenNoExplicitPortConfigured| - [TcpServerConnection] hz.dreamy_blackburn.IO.thread-in-1 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Initialized new cluster connection between /0.0.0.0:40592 and /127.0.0.1:5702
09:01:33,444  INFO |test_whenNoExplicitPortConfigured| - [TcpIpJoiner] hz.dreamy_blackburn.cached.thread-3 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] [127.0.0.1]:5703 is added to the blacklist.
09:01:33,444  WARN |test_whenNoExplicitPortConfigured| - [ClusterJoinManager] hz.stupefied_blackburn.priority-generic-operation.thread-0 - [127.0.0.1]:5702 [dev] [5.0-SNAPSHOT] New join request has been received from an existing endpoint Member [127.0.0.1]:5701 - 70b81027-7809-49bb-8895-e4adf0fc7475. Removing old member and processing join request...
09:01:33,444  WARN |test_whenNoExplicitPortConfigured| - [MembershipManager] hz.stupefied_blackburn.priority-generic-operation.thread-0 - [127.0.0.1]:5702 [dev] [5.0-SNAPSHOT] Member [127.0.0.1]:5701 - 70b81027-7809-49bb-8895-e4adf0fc7475 is suspected to be dead for reason: New join request has been received from an existing endpoint Member [127.0.0.1]:5701 - 70b81027-7809-49bb-8895-e4adf0fc7475. Removing old member and processing join request...
09:01:33,444  INFO |test_whenNoExplicitPortConfigured| - [MembershipManager] hz.stupefied_blackburn.priority-generic-operation.thread-0 - [127.0.0.1]:5702 [dev] [5.0-SNAPSHOT] Starting mastership claim process...
09:01:33,444  INFO |test_whenNoExplicitPortConfigured| - [MembershipManager] hz.stupefied_blackburn.priority-generic-operation.thread-0 - [127.0.0.1]:5702 [dev] [5.0-SNAPSHOT] Local MembersView{version=2, members=[MemberInfo{address=[127.0.0.1]:5701, uuid=70b81027-7809-49bb-8895-e4adf0fc7475, liteMember=false, memberListJoinVersion=1}, MemberInfo{address=[127.0.0.1]:5702, uuid=bce25553-6b6d-4d14-a003-674b088f9bc6, liteMember=false, memberListJoinVersion=2}]} with suspected members: [Member [127.0.0.1]:5701 - 70b81027-7809-49bb-8895-e4adf0fc7475] and initial addresses to ask: []
09:01:33,444  INFO |test_whenNoExplicitPortConfigured| - [TcpServerConnection] hz.stupefied_blackburn.cached.thread-2 - [127.0.0.1]:5702 [dev] [5.0-SNAPSHOT] Connection[id=2, /127.0.0.1:5702->/127.0.0.1:40592, qualifier=null, endpoint=[127.0.0.1]:5701, alive=false, connectionType=MEMBER, planeIndex=0] closed. Reason: Member left event received from master
09:01:33,444  INFO |test_whenNoExplicitPortConfigured| - [TcpServerConnection] hz.dreamy_blackburn.IO.thread-in-1 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Connection[id=1, /0.0.0.0:40592->/127.0.0.1:5702, qualifier=null, endpoint=[127.0.0.1]:5702, alive=false, connectionType=MEMBER, planeIndex=0] closed. Reason: Connection closed by the other side
09:01:33,444  INFO |test_whenNoExplicitPortConfigured| - [ClusterService] hz.stupefied_blackburn.cached.thread-2 - [127.0.0.1]:5702 [dev] [5.0-SNAPSHOT] 

Members {size:1, ver:3} [
	Member [127.0.0.1]:5702 - bce25553-6b6d-4d14-a003-674b088f9bc6 this
]

09:01:33,444  INFO |test_whenNoExplicitPortConfigured| - [TransactionManagerService] hz.stupefied_blackburn.cached.thread-4 - [127.0.0.1]:5702 [dev] [5.0-SNAPSHOT] Committing/rolling-back live transactions of [127.0.0.1]:5701, UUID: 70b81027-7809-49bb-8895-e4adf0fc7475
09:01:33,444  INFO |test_whenNoExplicitPortConfigured| - [MembershipManager] hz.stupefied_blackburn.cached.thread-2 - [127.0.0.1]:5702 [dev] [5.0-SNAPSHOT] Mastership is claimed with: MembersView{version=3, members=[MemberInfo{address=[127.0.0.1]:5702, uuid=bce25553-6b6d-4d14-a003-674b088f9bc6, liteMember=false, memberListJoinVersion=2}]}
09:01:33,445  INFO |test_whenNoExplicitPortConfigured| - [TcpIpJoiner] hz.dreamy_blackburn.IO.thread-in-1 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] [127.0.0.1]:5702 is added to the blacklist.
09:01:33,644  INFO |test_whenNoExplicitPortConfigured| - [TcpIpJoiner] hz.dreamy_blackburn.cached.thread-3 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] [127.0.0.1]:5703 is added to the blacklist.
09:01:33,695  INFO |test_whenNoExplicitPortConfigured| - [InternalPartitionService] hz.stupefied_blackburn.migration - [127.0.0.1]:5702 [dev] [5.0-SNAPSHOT] Fetching partition tables from cluster to determine the most recent one... Local stamp: 0
09:01:33,695  INFO |test_whenNoExplicitPortConfigured| - [InternalPartitionService] hz.stupefied_blackburn.migration - [127.0.0.1]:5702 [dev] [5.0-SNAPSHOT] Most recent partition table is determined.
09:01:33,945  INFO |test_whenNoExplicitPortConfigured| - [TcpIpJoiner] hz.dreamy_blackburn.cached.thread-1 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] [127.0.0.1]:5703 is added to the blacklist.
09:01:34,344  INFO |test_whenNoExplicitPortConfigured| - [JetExtension] test_whenNoExplicitPortConfigured - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Jet extension is enabled after the cluster version upgrade.
09:01:34,344  INFO |test_whenNoExplicitPortConfigured| - [ClusterService] test_whenNoExplicitPortConfigured - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] 

Members {size:1, ver:1} [
	Member [127.0.0.1]:5701 - 4292310d-8a4d-4dbd-ba89-5ca9f97b03bd this
]

09:01:34,344  INFO |test_whenNoExplicitPortConfigured| - [JetExtension] test_whenNoExplicitPortConfigured - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Jet extension is enabled
09:01:34,344 DEBUG |test_whenNoExplicitPortConfigured| - [JobCoordinationService] hz.dreamy_blackburn.cached.thread-5 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Not starting jobs because partitions are not yet initialized.
09:01:34,344 DEBUG |test_whenNoExplicitPortConfigured| - [JobCoordinationService] hz.dreamy_blackburn.cached.thread-6 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Not starting jobs because partitions are not yet initialized.
09:01:34,345  INFO |test_whenNoExplicitPortConfigured| - [LifecycleService] test_whenNoExplicitPortConfigured - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] [127.0.0.1]:5701 is STARTED
09:01:34,364  INFO |test_whenNoExplicitPortConfigured| - [HealthMonitor] hz.dreamy_blackburn.HealthMonitor - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] processors=72, physical.memory.total=377.6G, physical.memory.free=151.2G, swap.space.total=4.0G, swap.space.free=2.2G, heap.memory.used=2.9G, heap.memory.free=1.1G, heap.memory.total=4.0G, heap.memory.max=4.0G, heap.memory.used/total=73.58%, heap.memory.used/max=73.58%, minor.gc.count=10599, minor.gc.time=275614ms, major.gc.count=67, major.gc.time=25273ms, load.process=2.85%, load.system=3.90%, load.systemAverage=5.81, thread.count=612, thread.peakCount=12615, cluster.timeDiff=0, event.q.size=0, executor.q.async.size=0, executor.q.client.size=0, executor.q.client.query.size=0, executor.q.client.blocking.size=0, executor.q.query.size=0, executor.q.scheduled.size=0, executor.q.io.size=0, executor.q.system.size=0, executor.q.operations.size=0, executor.q.priorityOperation.size=0, operations.completed.count=0, executor.q.mapLoad.size=0, executor.q.mapLoadAllKeys.size=0, executor.q.cluster.size=0, executor.q.response.size=0, operations.running.count=0, operations.pending.invocations.percentage=0.00%, operations.pending.invocations.count=0, proxy.count=0, clientEndpoint.count=0, connection.active.count=0, client.connection.count=0, connection.count=0
09:01:34,386  INFO |test_whenNoExplicitPortConfigured| - [LifecycleService] main - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] [127.0.0.1]:5701 is SHUTTING_DOWN
09:01:34,386  WARN |test_whenNoExplicitPortConfigured| - [Node] main - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Terminating forcefully...
09:01:34,386  INFO |test_whenNoExplicitPortConfigured| - [Node] main - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Shutting down connection manager...
09:01:34,386  INFO |test_whenNoExplicitPortConfigured| - [Node] main - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Shutting down node engine...
09:01:34,392  INFO |test_whenNoExplicitPortConfigured| - [NodeExtension] main - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Destroying node NodeExtension.
09:01:34,392  INFO |test_whenNoExplicitPortConfigured| - [Node] main - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Hazelcast Shutdown is completed in 6 ms.
09:01:34,392  INFO |test_whenNoExplicitPortConfigured| - [LifecycleService] main - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] [127.0.0.1]:5701 is SHUTDOWN
09:01:34,392  INFO |test_whenNoExplicitPortConfigured| - [LifecycleService] main - [127.0.0.1]:5702 [dev] [5.0-SNAPSHOT] [127.0.0.1]:5702 is SHUTTING_DOWN
09:01:34,392  WARN |test_whenNoExplicitPortConfigured| - [Node] main - [127.0.0.1]:5702 [dev] [5.0-SNAPSHOT] Terminating forcefully...
09:01:34,392  INFO |test_whenNoExplicitPortConfigured| - [Node] main - [127.0.0.1]:5702 [dev] [5.0-SNAPSHOT] Shutting down connection manager...
09:01:34,392  INFO |test_whenNoExplicitPortConfigured| - [Node] main - [127.0.0.1]:5702 [dev] [5.0-SNAPSHOT] Shutting down node engine...
09:01:34,398  INFO |test_whenNoExplicitPortConfigured| - [NodeExtension] main - [127.0.0.1]:5702 [dev] [5.0-SNAPSHOT] Destroying node NodeExtension.
09:01:34,398  INFO |test_whenNoExplicitPortConfigured| - [Node] main - [127.0.0.1]:5702 [dev] [5.0-SNAPSHOT] Hazelcast Shutdown is completed in 6 ms.
09:01:34,398  INFO |test_whenNoExplicitPortConfigured| - [LifecycleService] main - [127.0.0.1]:5702 [dev] [5.0-SNAPSHOT] [127.0.0.1]:5702 is SHUTDOWN
BuildInfo right after test_whenNoExplicitPortConfigured(com.hazelcast.cluster.TcpIpJoinTest): BuildInfo{version='5.0-SNAPSHOT', build='20210524', buildNumber=20210524, revision=c3c641f, enterprise=false, serializationVersion=1, jet=JetBuildInfo{version='5.0-SNAPSHOT', build='20210524', revision='c3c641f'}}
Hiccups measured while running test 'test_whenNoExplicitPortConfigured(com.hazelcast.cluster.TcpIpJoinTest):'
09:01:30, accumulated pauses: 43 ms, max pause: 16 ms, pauses over 1000 ms: 0
```

--- a/hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/ClusterServiceImpl.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/ClusterServiceImpl.java
@@ -73,6 +73,7 @@ import java.util.Properties;
 import java.util.Set;
 import java.util.UUID;
 import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.concurrent.locks.ReentrantLock;
 
 import static com.hazelcast.cluster.impl.MemberImpl.NA_MEMBER_LIST_JOIN_VERSION;
@@ -86,8 +87,6 @@ import static com.hazelcast.internal.util.Preconditions.checkFalse;
 import static com.hazelcast.internal.util.Preconditions.checkNotNull;
 import static com.hazelcast.internal.util.Preconditions.checkTrue;
 import static java.lang.String.format;
-import java.util.concurrent.CountDownLatch;
-import java.util.concurrent.atomic.AtomicReference;
 
 @SuppressWarnings({"checkstyle:methodcount", "checkstyle:classdataabstractioncoupling", "checkstyle:classfanoutcomplexity"})
 public class ClusterServiceImpl implements ClusterService, ConnectionListener, ManagedService,
@@ -116,21 +115,11 @@ public class ClusterServiceImpl implements ClusterService, ConnectionListener, M
     private final ClusterStateManager clusterStateManager;
     private final ClusterHeartbeatManager clusterHeartbeatManager;
     private final ReentrantLock lock = new ReentrantLock();
-    private final AtomicReference<JoinHolder> joined =
-            new AtomicReference<>(new JoinHolder(false));
+    private final AtomicBoolean joined = new AtomicBoolean(false);
 
     private volatile UUID clusterId;
     private volatile Address masterAddress;
     private volatile MemberImpl localMember;
-
-    private static class JoinHolder {
-        private final CountDownLatch latch = new CountDownLatch(1);
-        private final boolean isJoined;
-
-        JoinHolder(boolean isJoined) {
-            this.isJoined = isJoined;
-        }
-    }
 
     public ClusterServiceImpl(Node node, MemberImpl localMember) {
         this.node = node;
@@ -648,7 +637,6 @@ public class ClusterServiceImpl implements ClusterService, ConnectionListener, M
             logger.fine("Setting master address to " + master);
         }
         masterAddress = master;
-        joined.getAndUpdate(holder -> new JoinHolder(holder.isJoined)).latch.countDown();
     }
 
     @Override
@@ -678,12 +666,12 @@ public class ClusterServiceImpl implements ClusterService, ConnectionListener, M
     // should be called under lock
     void setJoined(boolean val) {
         assert lock.isHeldByCurrentThread() : "Called without holding cluster service lock!";
-        joined.getAndUpdate(holder -> new JoinHolder(val)).latch.countDown();
+        joined.set(val);
     }
 
     @Override
     public boolean isJoined() {
-        return joined.get().isJoined;
+        return joined.get();
     }
 
     @Probe(name = CLUSTER_METRIC_CLUSTER_SERVICE_SIZE)
@@ -1071,15 +1059,5 @@ public class ClusterServiceImpl implements ClusterService, ConnectionListener, M
     @Override
     public String toString() {
         return "ClusterService" + "{address=" + getThisAddress() + '}';
-    }
-
-    /**
-     *
-     * @param millis
-     * @return true is cluster has been joined, false if timed out
-     * @throws InterruptedException
-     */
-    public boolean blockOnJoin(long millis) throws InterruptedException {
-        return joined.get().latch.await(millis, TimeUnit.MILLISECONDS);
     }
 }

--- a/hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/MulticastJoiner.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/MulticastJoiner.java
@@ -96,7 +96,7 @@ public class MulticastJoiner extends AbstractJoiner {
             }
 
             try {
-                clusterService.blockOnJoin(JOIN_RETRY_INTERVAL);
+                Thread.sleep(JOIN_RETRY_INTERVAL);
             } catch (InterruptedException e) {
                 currentThread().interrupt();
             }

--- a/hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/TcpIpJoiner.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/TcpIpJoiner.java
@@ -26,7 +26,6 @@ import com.hazelcast.instance.impl.Node;
 import com.hazelcast.internal.cluster.impl.SplitBrainJoinMessage.SplitBrainMergeCheckResult;
 import com.hazelcast.internal.cluster.impl.operations.JoinMastershipClaimOp;
 import com.hazelcast.internal.nio.Connection;
-import com.hazelcast.internal.server.ServerConnectionManager;
 import com.hazelcast.internal.util.AddressUtil;
 import com.hazelcast.internal.util.AddressUtil.AddressMatcher;
 import com.hazelcast.internal.util.AddressUtil.InvalidAddressException;
@@ -113,11 +112,10 @@ public class TcpIpJoiner extends AbstractJoiner {
             long joinStartTime = Clock.currentTimeMillis();
             Connection connection;
             while (shouldRetry() && (Clock.currentTimeMillis() - joinStartTime < maxJoinMillis)) {
-                ServerConnectionManager connectionManager = node.getServer().getConnectionManager(MEMBER);
-                connection = connectionManager.getOrConnect(targetAddress);
+                connection = node.getServer().getConnectionManager(MEMBER).getOrConnect(targetAddress);
                 if (connection == null) {
                     //noinspection BusyWait
-                    connectionManager.blockOnConnect(targetAddress, JOIN_RETRY_WAIT_TIME, 0);
+                    Thread.sleep(JOIN_RETRY_WAIT_TIME);
                     continue;
                 }
                 if (logger.isFineEnabled()) {
@@ -125,9 +123,7 @@ public class TcpIpJoiner extends AbstractJoiner {
                 }
                 clusterJoinManager.sendJoinRequest(targetAddress);
                 //noinspection BusyWait
-                if (!clusterService.isJoined()) {
-                    clusterService.blockOnJoin(JOIN_RETRY_WAIT_TIME);
-                }
+                Thread.sleep(JOIN_RETRY_WAIT_TIME);
             }
         } catch (final Exception e) {
             logger.warning(e);
@@ -263,7 +259,7 @@ public class TcpIpJoiner extends AbstractJoiner {
             }
 
             if (!clusterService.isJoined()) {
-                clusterService.blockOnJoin(JOIN_RETRY_WAIT_TIME);
+                Thread.sleep(JOIN_RETRY_WAIT_TIME);
             }
         }
     }

--- a/hazelcast/src/main/java/com/hazelcast/internal/server/ServerConnectionManager.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/server/ServerConnectionManager.java
@@ -24,8 +24,6 @@ import com.hazelcast.internal.nio.Packet;
 import javax.annotation.Nonnull;
 import javax.annotation.Nullable;
 import java.util.Collection;
-import java.util.concurrent.TimeUnit;
-import java.util.concurrent.locks.LockSupport;
 import java.util.function.Consumer;
 import java.util.function.Predicate;
 
@@ -215,18 +213,4 @@ public interface ServerConnectionManager
      * @return the Server.
      */
     Server getServer();
-
-    /**
-     * blocks the caller thread until a connection is established (or failed)
-     * or the time runs out. Callers must ensure a connection is established after this method returns {@code true}.
-     * @param address
-     * @param millis
-     * @param streamId
-     * @return true if connected successfully, false if timed out
-     * @throws java.lang.InterruptedException
-     */
-    default boolean blockOnConnect(Address address, long millis, int streamId) throws InterruptedException {
-        LockSupport.parkNanos(TimeUnit.MILLISECONDS.toNanos(millis));
-        return false;
-    }
 }

--- a/hazelcast/src/main/java/com/hazelcast/internal/server/ServerContext.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/server/ServerContext.java
@@ -37,7 +37,6 @@ import java.net.Socket;
 import java.util.Collection;
 import java.util.Map;
 import java.util.UUID;
-import java.util.concurrent.Future;
 
 /**
  * Contains many of the dependencies passed to the {@link Server}.
@@ -102,7 +101,7 @@ public interface ServerContext {
 
     void onDisconnect(Address endpoint, Throwable cause);
 
-    Future<Void> submitAsync(Runnable runnable);
+    void executeAsync(Runnable runnable);
 
     EventService getEventService();
 

--- a/hazelcast/src/main/java/com/hazelcast/internal/server/tcp/TcpServerAcceptor.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/server/tcp/TcpServerAcceptor.java
@@ -296,7 +296,7 @@ public class TcpServerAcceptor implements DynamicMetricsProvider {
                 .addParameter("remoteAddress", socketChannel.getRemoteAddress())
                 .log();
             if (serverContext.isSocketInterceptorEnabled(qualifier)) {
-                serverContext.submitAsync(() -> newConnection0(connectionManager, channel));
+                serverContext.executeAsync(() -> newConnection0(connectionManager, channel));
             } else {
                 newConnection0(connectionManager, channel);
             }

--- a/hazelcast/src/main/java/com/hazelcast/internal/server/tcp/TcpServerConnectionManager.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/server/tcp/TcpServerConnectionManager.java
@@ -32,7 +32,6 @@ import com.hazelcast.internal.nio.Packet;
 import com.hazelcast.internal.server.NetworkStats;
 import com.hazelcast.internal.server.ServerConnection;
 import com.hazelcast.internal.server.ServerContext;
-import com.hazelcast.internal.util.ExceptionUtil;
 import com.hazelcast.internal.util.executor.StripedRunnable;
 
 import java.io.IOException;
@@ -40,11 +39,6 @@ import java.nio.channels.SocketChannel;
 import java.util.Collection;
 import java.util.Map;
 import java.util.Set;
-import java.util.concurrent.ExecutionException;
-import java.util.concurrent.Future;
-import java.util.concurrent.TimeUnit;
-import java.util.concurrent.TimeoutException;
-import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.concurrent.atomic.AtomicInteger;
 import java.util.function.Consumer;
 import java.util.function.Function;
@@ -127,15 +121,11 @@ public class TcpServerConnectionManager extends TcpServerConnectionManagerBase
         Plane plane = getPlane(streamId);
         TcpServerConnection connection = plane.getConnection(address);
         if (connection == null && server.isLive()) {
-            final AtomicBoolean isNotYetInProgress = new AtomicBoolean();
-            plane.addConnectionInProgressIfAbsent(address, key -> {
-                isNotYetInProgress.set(true);
-                return connector.asyncConnect(address, silent, plane.index);
-            });
-            if (isNotYetInProgress.get()) {
+            if (plane.addConnectionInProgress(address)) {
                 if (logger.isFineEnabled()) {
                     logger.fine("Connection to: " + address + " streamId:" + streamId + " is not yet progress");
                 }
+                connector.asyncConnect(address, silent, plane.index);
             } else {
                 if (logger.isFineEnabled()) {
                     logger.fine("Connection to: " + address + " streamId:" + streamId + " is already in progress");
@@ -352,24 +342,6 @@ public class TcpServerConnectionManager extends TcpServerConnectionManagerBase
         if (endpointConfig == null) {
             context.collect(rootDescriptor.copy(), TCP_METRIC_CLIENT_COUNT, MANDATORY, COUNT, clientCount);
             context.collect(rootDescriptor.copy(), TCP_METRIC_TEXT_COUNT, MANDATORY, COUNT, textCount);
-        }
-    }
-
-    @Override
-    public boolean blockOnConnect(Address address, long millis, int streamId) throws InterruptedException {
-        Plane plane = getPlane(streamId);
-        try {
-            Future<Void> future = plane.getconnectionInProgress(address);
-            if (future != null) {
-                future.get(millis, TimeUnit.MILLISECONDS);
-                return future.isDone();
-            } else {
-                // connection-in-progress has come and gone,
-                // so we are not timed out here
-                return true;
-            }
-        } catch (ExecutionException | TimeoutException ex) {
-            throw ExceptionUtil.rethrow(ex);
         }
     }
 }

--- a/hazelcast/src/main/java/com/hazelcast/internal/server/tcp/TcpServerConnectionManagerBase.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/server/tcp/TcpServerConnectionManagerBase.java
@@ -40,12 +40,10 @@ import java.util.Map;
 import java.util.Set;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.CopyOnWriteArraySet;
-import java.util.concurrent.Future;
 import java.util.concurrent.RejectedExecutionException;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicLong;
 import java.util.function.Consumer;
-import java.util.function.Function;
 
 import static com.hazelcast.internal.metrics.MetricDescriptorConstants.TCP_METRIC_ENDPOINT_MANAGER_ACCEPTED_SOCKET_COUNT;
 import static com.hazelcast.internal.metrics.MetricDescriptorConstants.TCP_METRIC_ENDPOINT_MANAGER_ACTIVE_COUNT;
@@ -123,8 +121,8 @@ abstract class TcpServerConnectionManagerBase implements ServerConnectionManager
         final ConcurrentHashMap<Address, TcpServerConnectionErrorHandler> errorHandlers = new ConcurrentHashMap<>(100);
         final int index;
 
-        private final Map<Address, Future<Void>> connectionsInProgress = new ConcurrentHashMap<>();
         private final ConcurrentHashMap<Address, TcpServerConnection> connectionMap = new ConcurrentHashMap<>(100);
+        private final Set<Address> connectionsInProgress = newSetFromMap(new ConcurrentHashMap<>());
 
         Plane(int index) {
             this.index = index;
@@ -186,22 +184,15 @@ abstract class TcpServerConnectionManagerBase implements ServerConnectionManager
         }
 
         public boolean hasConnectionInProgress(Address address) {
-            return connectionsInProgress.containsKey(address);
+            return connectionsInProgress.contains(address);
         }
 
-        public Future<Void> getconnectionInProgress(Address address) {
-            return connectionsInProgress.get(address);
-        }
-
-        public void addConnectionInProgressIfAbsent(
-                Address address,
-                Function<? super Address, ? extends Future<Void>> mappingFn
-        ) {
-            connectionsInProgress.computeIfAbsent(address, mappingFn);
+        public boolean addConnectionInProgress(Address address) {
+            return connectionsInProgress.add(address);
         }
 
         public boolean removeConnectionInProgress(Address address) {
-            return connectionsInProgress.remove(address) != null;
+            return connectionsInProgress.remove(address);
         }
 
         public void clearConnectionsInProgress() {

--- a/hazelcast/src/main/java/com/hazelcast/internal/server/tcp/TcpServerConnector.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/server/tcp/TcpServerConnector.java
@@ -34,7 +34,6 @@ import java.net.UnknownHostException;
 import java.nio.channels.SocketChannel;
 import java.util.Collection;
 import java.util.LinkedList;
-import java.util.concurrent.Future;
 import java.util.logging.Level;
 
 import static com.hazelcast.spi.properties.ClusterProperty.SOCKET_CLIENT_BIND;
@@ -73,9 +72,9 @@ class TcpServerConnector {
         this.planeCount = connectionManager.planeCount;
     }
 
-    Future<Void> asyncConnect(Address address, boolean silent, int planeIndex) {
+    void asyncConnect(Address address, boolean silent, int planeIndex) {
         serverContext.shouldConnectTo(address);
-        return serverContext.submitAsync(new ConnectTask(address, silent, planeIndex));
+        serverContext.executeAsync(new ConnectTask(address, silent, planeIndex));
     }
 
     private boolean useAnyOutboundPort() {

--- a/hazelcast/src/main/java/com/hazelcast/internal/server/tcp/TcpServerContext.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/server/tcp/TcpServerContext.java
@@ -54,7 +54,6 @@ import java.util.Collection;
 import java.util.Collections;
 import java.util.Map;
 import java.util.UUID;
-import java.util.concurrent.Future;
 import java.util.concurrent.TimeUnit;
 
 import static com.hazelcast.instance.EndpointQualifier.MEMBER;
@@ -266,9 +265,8 @@ public class TcpServerContext implements ServerContext {
     }
 
     @Override
-    @SuppressWarnings("unchecked")
-    public Future<Void> submitAsync(final Runnable runnable) {
-        return (Future<Void>) nodeEngine.getExecutionService().submit(ExecutionService.IO_EXECUTOR, runnable);
+    public void executeAsync(final Runnable runnable) {
+        nodeEngine.getExecutionService().execute(ExecutionService.IO_EXECUTOR, runnable);
     }
 
     @Override

--- a/hazelcast/src/test/java/com/hazelcast/client/ClientRegressionWithRealNetworkTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/client/ClientRegressionWithRealNetworkTest.java
@@ -247,7 +247,7 @@ public class ClientRegressionWithRealNetworkTest extends ClientTestSupport {
 
         warmUpPartitions(instance1, instance2);
         String keyOwnedBy2 = generateKeyOwnedBy(instance2);
-        makeSureConnectedToServers(client, 2);
+
 
         IMap<Object, Object> clientMap = client.getMap("test");
 

--- a/hazelcast/src/test/java/com/hazelcast/internal/partition/impl/FrozenPartitionTableTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/internal/partition/impl/FrozenPartitionTableTest.java
@@ -258,7 +258,6 @@ public class FrozenPartitionTableTest extends HazelcastTestSupport {
         newHazelcastInstance(initOrCreateConfig(new Config()),
                 randomName(), new StaticMemberNodeContext(factory, member4.getUuid(), member3.getAddress()));
         assertClusterSizeEventually(3, hz1, hz2);
-        waitAllForSafeState(hz1, hz2);
 
         OperationServiceImpl operationService = getOperationService(hz1);
         operationService.invokeOnPartition(null, new NonRetryablePartitionOperation(), member3PartitionId).join();

--- a/hazelcast/src/test/java/com/hazelcast/internal/server/MockServerContext.java
+++ b/hazelcast/src/test/java/com/hazelcast/internal/server/MockServerContext.java
@@ -55,8 +55,6 @@ import java.util.Map;
 import java.util.Properties;
 import java.util.UUID;
 import java.util.concurrent.CompletableFuture;
-import java.util.concurrent.Future;
-import java.util.concurrent.FutureTask;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.function.Consumer;
 
@@ -198,16 +196,14 @@ public class MockServerContext implements ServerContext {
     }
 
     @Override
-    public Future<Void> submitAsync(final Runnable runnable) {
-        FutureTask<Void> future = new FutureTask<>(() -> {
+    public void executeAsync(final Runnable runnable) {
+        new Thread(() -> {
             try {
                 runnable.run();
             } catch (Throwable t) {
                 logger.severe(t);
             }
-        }, null);
-        new Thread(() -> future.run()).start();
-        return future;
+        }).start();
     }
 
     @Override

--- a/hazelcast/src/test/java/com/hazelcast/test/mocknetwork/MockJoiner.java
+++ b/hazelcast/src/test/java/com/hazelcast/test/mocknetwork/MockJoiner.java
@@ -83,7 +83,7 @@ class MockJoiner extends AbstractJoiner {
                 break;
             }
             try {
-                clusterService.blockOnJoin(500);
+                Thread.sleep(500);
             } catch (InterruptedException e) {
                 e.printStackTrace();
                 break;

